File: moderate/justify.lua
function justi(s)
  if #s == 80 then
    print(s)
  else
    local r = {}
    for i in s:gmatch("%S+") do
      r[#r+1] = i
    end
    local l = 80 - #table.concat(r)
    for i = 1, l%(#r-1) do
      r[i] = r[i] .. " "
    end
    print(table.concat(r, string.rep(" ", math.floor(l/(#r-1)))))
  end
end

function maxwords(s)
  local t = (#s > 81) and s:sub(1,81) or s
  return t:find(" %S*$") - 1
end

function justify(s)
  local t = s:match"^%s*(.*)"
  if #t <= 80 then
    print(t)
  else
    local m = maxwords(t)
    justi(t:sub(1, m))
    return justify(t:sub(m+1))
  end
end

for line in io.lines(arg[1]) do
  if #line > 80 then
    justify(line)
  else
    print(line)
  end
end

File: moderate/justify.hs
import System.Environment (getArgs)

spaces       :: [String] -> Int -> Int -> [String]
spaces xs 0 y = xs
spaces xs x y | x >= y    = spaces (map (++ " ") (init xs) ++ [last xs]) (x-y) y
              | otherwise = map (++ " ") (take x xs) ++ drop x xs

justi :: String -> String
justi xs | length xs == 80 = xs
         | otherwise       = concat (spaces ws l (length ws - 1))
         where ws = words xs
               l  = 80 - sum (map length ws)

maxwords :: String -> Int
maxwords [] = -1
maxwords xs | length xs > 81 = maxwords (take 81 xs)
            | last xs == ' ' = length xs - 1
            | otherwise      = maxwords (init xs)

justif   :: String -> [String]
justif xs | head xs == ' '  = justif (tail xs)
          | length xs <= 80 = [xs]
          | otherwise       = justi (take m xs) : justif (drop m xs)
          where m = maxwords xs

justify   :: String -> String
justify xs | length xs <= 80 = xs
           | otherwise       = init (unlines (justif xs))

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map justify $ lines input
